### ISSUE TITLE
problem with libspotify gem on armv6?

### ISSUE THREAD

**Issue by bastiaanterhorst** [Burgestrand/libspotify#1]
Hi,
I'm running into a bit of a problem trying to get libspotify and hallon working on a Raspberry Pi (a small ARM-based computer board). The gem installs just fine, but running a simple program (like the examples provided in the Hallon readme) segfault.

Running this example:

``` ruby
#!/usr/bin/ruby -w

require 'rubygems'
require 'bundler'

require 'hallon'
require 'hallon-openal'

session = Hallon::Session.initialize IO.read('./spotify_appkey.key')
session.login!('secret', 'secret') #obviously I had real credentials here

track = Hallon::Track.new("spotify:track:1ZPsdTkzhDeHjA5c2Rnt2I")
track.load

player = Hallon::Player.new(Hallon::OpenAL)
player.play!(track)
```

Yields this error:

``` bash
pi@raspberrypi ~/spot $ ./spot.rb 
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:31: warning: method redefined; discarding old visit_String
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:4: warning: previous definition of visit_String was here
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:39: warning: method redefined; discarding old register
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:12: warning: previous definition of register was here
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:43: warning: method redefined; discarding old format_time
/usr/lib/ruby/1.9.1/rubygems/psych_tree.rb:16: warning: previous definition of format_time was here
/usr/lib/ruby/1.9.1/rubygems/syck_hack.rb:115: warning: method redefined; discarding old to_s
/usr/lib/ruby/1.9.1/rubygems/syck_hack.rb:44: warning: previous definition of to_s was here
/var/lib/gems/1.9.1/gems/hallon-0.18.1/lib/hallon/session.rb:49: warning: instance variable @__instance__ not initialized
/var/lib/gems/1.9.1/gems/spotify-12.3.0/lib/spotify.rb:99: [BUG] Segmentation fault
ruby 1.9.3p194 (2012-04-20 revision 35410) [arm-linux-eabi]

-- Control frame information -----------------------------------------------
c:0022 p:---- s:0074 b:0074 l:000073 d:000073 CFUNC  :session_process_events
c:0021 p:0020 s:0069 b:0069 l:000063 d:000068 BLOCK  /var/lib/gems/1.9.1/gems/spotify-12.3.0/lib/spotify.rb:99
c:0020 p:0019 s:0067 b:0067 l:000066 d:000066 METHOD /usr/lib/ruby/1.9.1/monitor.rb:211
c:0019 p:0013 s:0064 b:0064 l:000063 d:000063 METHOD /var/lib/gems/1.9.1/gems/spotify-12.3.0/lib/spotify.rb:98
c:0018 p:---- s:0058 b:0058 l:000057 d:000057 FINISH
c:0017 p:0025 s:0056 b:0056 l:000044 d:000055 BLOCK  /var/lib/gems/1.9.1/gems/hallon-0.18.1/lib/hallon/session.rb:168
c:0016 p:---- s:0053 b:0053 SEGV received in SEGV handler
Aborted
```

If it helps, this is on a Raspberry Pi board running "Soft-Float Raspbian Wheezy". I'm running Ruby ruby 1.9.3p194. I installed the libspotify gem, but after I encountered this issue I also compiled C libspotify by hand following your guide. There was no difference (although your Ruby libspotify might favor it's own binary libspotify over the one I compiled – I did not dive into the source to check this).

Do you have any idea what's going on or hints as to where I might look to get this working?

**Comment by Burgestrand** [comment]
As far as libspotify on the raspberry pi I’ve seen _a lot_ of people having issues with it on twitter. I don’t have one myself, so I’m not able to experiment with it either.

Given the interesting platform of the raspberry pi (Hallon π!), the libspotify gem will not install a binary distribution of libspotify. As a result, the `LIBSPOTIFY_BIN` path exported by this gem simply don’t point to a file at all, and the Spotify gem will try to find libspotify by going through the paths searched by the `dlopen` C function (usually LD_LIBRARY_PATH and something more).

So, if I am guessing right, the libspotify gem contains nothing on your platform. You should be able to check this by writing this line in a terminal: `ruby -rlibspotify -e 'p File.exists?(LIBSPOTIFY_BIN)'`

I should be able to publish a version for the raspberry pi as well. If it’s not too much trouble, could you also paste the output of `ruby -rubygems -e 'p Gem::Platform.local.to_s'`? It’ll (possibly) tell me what platform I should tell rubygems that the binary distribution supports.

**EDIT**: Actually, if the check about LIBSPOTIFY_BIN returns true that’ll probably be why it segfaults.

**Comment by bastiaanterhorst** [comment]
Thanks for your answer, and please excuse the delay in my reply.

The first command returns false:

```
pi@raspberrypi ~ $ ruby -rlibspotify -e 'p File.exists?(LIBSPOTIFY_BIN)'
false
```

The second command "arm-linux":

```
pi@raspberrypi ~ $ ruby -rubygems -e 'p Gem::Platform.local.to_s'
"arm-linux"
```

Hope this helps!

**Comment by Burgestrand** [comment]
Hm, still no indication to what your issue is about, and frankly I don’t even know if it is _supposed_ to work. Here’s what I have been able to figure out.
1. You are not using the binary distribution of libspotify from the libspotify gem, because it has no binary distribution for the R-pi. This means that the libspotify gem (this gem) is not the cause of your issue. It also means that the [spotify gem](https://rubygems.org/gems/spotify) will be using your manual installation of libspotify.
2. You can run the example, even if it crashes, and it doesn’t crash because of an issue with the libspotify version or being unable to load libspotify. Therefore we can assume you have the correct version of libspotify installed from your manual installation. This is also a good thing, meaning it’s not a version incompatibility with libspotify.
3. Issues related to libspotify on R-pi are usually blamed on hard floats, but since you are using soft floats I’m not sure if your issue is your OS distribution. I do know that pyspotify tells you to use e.g. Arch linux instead of the Raspbian image when using libspotify. This might be the issue, might not. I would try using another OS distribution.
4. Given that it crashes on `process_events` I can also assume it is crashing _after_ creating the session, most likely it crashes when trying to log in.

Given the above points, there’s a few possible reasons to why it does not work for you:
1. Either libspotify simply does not work on your current raspbian OS, and you’ll need to run another OS.
2. The spotify gem is incorrectly configured on your platform. You can test this by running `ruby -rspotify -e 'p Spotify.linux?'`, if it prints out _true_ then the spotify gem should be correctly configured for your platform.
3. It’s possible that Hallon could do things that won’t work with the R-pi. If this is the case my main guess would be that the callback support is breaking things in this case. You could try logging in and retrieving minimal track information with this script: https://gist.github.com/3433448. If it works, then there is a bug in Hallon or in the callback support.

In other words I’ve not yet tracked down the exact cause of your issue, and it’s just a bunch of guesses. I completely understand if this is asking too much.

There is always Mopidy (http://www.mopidy.com/) if you want to play music on your R-pi. It’s written in python and also uses libspotify. I believe the Mopidy guys are working on R-pi support, and that people have gotten it to work.

**Comment by bastiaanterhorst** [comment]
Thanks for the extensive reply! Tomorrow I will have time to investigate. I'll let you know how it goes.

**Comment by bastiaanterhorst** [comment]
Your gist returns the following:

```

pi@raspberrypi ~/spot $ ./test.rb 
Creating session.
Created! Logging in.
Log in requested. Waiting forever until logged in.
Logged in as bastiaanterhorst. Requesting track information.
Track created. Waiting forever until track is loaded.
Track loaded: Trick Me.

From: ./test.rb @ line 63 :

/var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/scanner.rb:309: [BUG] Segmentation fault
ruby 1.9.3p194 (2012-04-20 revision 35410) [arm-linux-eabi]

-- Control frame information -----------------------------------------------
c:0041 p:---- s:0193 b:0193 l:000192 d:000192 CFUNC  :set_backtrace
c:0040 p:---- s:0187 b:0187 l:000186 d:000186 CFUNC  :backtrace
c:0039 p:0232 s:0185 b:0185 l:000184 d:000184 METHOD /var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/scanner.rb:282
c:0038 p:0086 s:0177 b:0177 l:000173 d:000176 BLOCK  /var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/scanner.rb:200
c:0037 p:0207 s:0174 b:0174 l:000173 d:000173 before_session hook failed: CodeRay::Scanners::Scanner::ScanError: 

***ERROR in scanner.rb:200:in `rescue in tokenize': regexp preprocess failed: invalid multibyte character (after 0 tokens)

tokens:


current line: 1  column: 6  pos: 5
matched: nil  state: "Error in CodeRay::Scanners::Ruby#scan_tokens, initial state was: :initial"
bol? = false,  eos? = false

surrounding code:
"r until track is loaded.\""  ~~  "\"Track created. Waiting foreve"


***ERROR***

/var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/scanners/ruby.rb:157:in `scan'
(see _pry_.hooks.errors to debug)
METHOD /var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/scanner.rb:196
c:0036 p:0094 s:0167 b:0167 l:000166 d:000166 METHOD /var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/encoder.rb:85
c:0035 p:---- s:0156 b:0156 l:000155 d:000155 CFUNC  :===
c:0034 p:0047 s:0152 b:0152 l:000148 d:000151 BLOCK 
c:0033 p:0022 s:0149 b:0149 l:000148 d:000148 METHOD /var/lib/gems/1.9.1/gems/coderay-1.0.7/lib/coderay/tokens_proxy.rb:32
c:0032 p:---- s:0143 b:0143 l:000142 SEGV received in SEGV handler
Aborted
```

So I guess there's a problem with Raspbian. I'll give Arch a try.
Thanks for all your info!

**Comment by Burgestrand** [comment]
It might be the issue with raspbian, but I find this interesting. The backtrace shows an error, but this time it comes from coderay, which is part of pry and not part of any of my spotify gems. It is interesting because: why would that be causing a segfault? It is a _raise_ call in nothing but ruby: https://github.com/rubychan/coderay/blob/v1.0.7/lib/coderay/scanner.rb#L282. Most likely something to do with raspbian or your current version of ruby.

Apart from that, I wonder if the code works if you decide to remove the last line (about `binding.pry`), and require the script from IRB. If it does, your initial issue might still come from Hallon. If not, we’ll be more or less sure that there is something else wrong here.

Anyhow, I understand you might’ve installed arch or even given up on this. You don’t need to do anything. I’m mainly posting this comment for myself and when (or if) I buy myself a raspberry pi for testing libspotify out on.

**Comment by bastiaanterhorst** [comment]
I have not given up! :)

And I also havent had time to try Arch. So when I get a chance (sometime next week) I'll try to remove Pry and see how that goes.

**Comment by Burgestrand** [comment]
Very resilient. :)

I’ve updated the gist to no longer rely on pry. It uses regular old IRB and global variables ($session, $link, $track) instead. I’ll see if I can throw in some code that lets you turn on and off basic callback support easily.

**Comment by bastiaanterhorst** [comment]
I've tried the updated gist, and it runs without errors!

```
pi@raspberrypi ~/spot $ ./test2.rb 
Creating session.
Created! Logging in.
Log in requested. Waiting forever until logged in.
Logged in as bastiaanterhorst. Requesting track information.
Track created. Waiting forever until track is loaded.
Track loaded: Trick Me.
irb(main):001:0> 

```

**Comment by bastiaanterhorst** [comment]
It would seem then, that something in Hallon (or something Hallon builds upon) is causing the crash?

For my needs, Mopidy will not do. I'm planning to write my own (simple) spotify client with a custom hardware interface. Your Hallon library seemed a perfect fit, since I'm already familiar with Ruby.

I would really appreciate any pointers. Should I give Arch a shot?

Thanks,
Bastiaan

**Comment by Burgestrand** [comment]
Great, thanks! I’ve not yet added callback support to the gist.

So far, you’ve managed to make sure that libspotify’s basic functions work fine. Like you say, it’s way more likely now that something part of Hallon is causing the crash. My current guess is that it has something to do with callback support in Hallon.

I’m at work at the moment, so I can’t really expand the gist to support callbacks until I get home. That’ll be about five hours from now.

But, like you’ve noticed, I have not yet understood the root cause, which makes it tricky to provide any helpful insight in how to work around it either. Given that libspotify appears to work, I would say it is improbable (but not impossible) that changing your OS to Arch would work, since it depended on libspotify being the culprit.

I’m very interested to see this start working as well, so I’ll be sticking to this issue either way. I’ll be ordering myself an R-Pi tonight so I might be able to start debugging even more fiercely. :)

**Comment by Burgestrand** [comment]
Oh, by the way, depending on how much of the API you’ll be needing it might not be crazy to ignore Hallon completely and instead just use the spotify gem: http://github.com/Burgestrand/spotify — it appears to work fine in your case. It’s a bit more low level, but frankly speaking it’s not extremely far from what Hallon offers. If you only use a subset of the API, it might not be that painful. Of course, if you have any questions or issues on developing with that gem I’m around to ask questions. :)

**Comment by Burgestrand** [comment]
Alright, I’ve added some basic callback usage to the gist, if you still want to try it out. It will now print a log message when libspotify asks it to, as well on log in.

I did not add logout to the gist, since it appears there’s a tricky bug in there that others have encountered with libspotify as well. If I added that, it might cloud the actual errors happening.

**Comment by bastiaanterhorst** [comment]
It runs without errors:

```
I, [2012-08-31T13:54:04.525782 #2072]  INFO -- : Creating session.
I, [2012-08-31T13:54:04.542222 #2072]  INFO -- : Created! Logging in.
I, [2012-08-31T13:54:05.247312 #2072]  INFO -- : Log in requested. Waiting forever until logged in.
I, [2012-08-31T13:54:05.251727 #2072]  INFO -- session (log message): 13:54:05.251 I [user_cache:135] UserCache::initiateGetUsers() will query for 1 users

I, [2012-08-31T13:54:05.253885 #2072]  INFO -- session (log message): 13:54:05.246 I [ap:1752] Connecting to AP ap.spotify.com:4070

I, [2012-08-31T13:54:05.272353 #2072]  INFO -- session (logged in): [OK] No error (0)
I, [2012-08-31T13:54:05.341874 #2072]  INFO -- session (log message): 13:54:05.340 I [ap:1226] Connected to AP: 78.31.8.51:4070

I, [2012-08-31T13:54:06.425084 #2072]  INFO -- : Logged in as bastiaanterhorst. Requesting track information.
I, [2012-08-31T13:54:06.430836 #2072]  INFO -- : Track created. Waiting forever until track is loaded.
I, [2012-08-31T13:54:06.518486 #2072]  INFO -- : Track loaded: Trick Me.
```

So that leaves me clueless as to where the problems were coming from.

I will look into using libspotify directly, but the main reason I found Hallon so interesting was it's built-in support for playing audio. If I understand correctly, this is much harder to accomplish when using libspotify directly.

**Comment by Burgestrand** [comment]
Wow, it runs! That is great news!… and also kind of sad, since it means the bug is very most likely part of Hallon.

And yes, Hallon does help quite a lot with the audio processing thing. Most of the help comes from the actual audio driver, however, which is the piece of code that actually _plays_ the audio. This is the part that is the trickiest to replace, and is probably the part of Hallon that has the most problems when using in raw form at the moment.

I’ll keep debugging this when I get my raspberry, but until then I don’t have anything more that you could try that would help me finding the root cause of this issue.

I don’t have much time to write a lengthy reply right now, but tonight I can work on writing a bare-bones example of playing audio with raw libspotify, but even when that is done it will be the minimum amount of work that needs to be done, and not a full application. However, it might still help you with your project, and it will help me clear up what needs to be fixed for the audio drivers apart from what I already know.

Thanks so much for the patience in trying all my code samples while I could not. I’ll try my best to repay the favor. :)

**Comment by bastiaanterhorst** [comment]
That would be great!
Meanwhile, I'm giving Arch a shot to see if it makes any difference.

**Comment by bastiaanterhorst** [comment]
It WORKS!
I switched to Arch, but I think the thing that made it work was to install the ARMv5 library instead of the ARMv6 one.

I can even play audio without problems :)

Thanks for all your assistance.

**Comment by Burgestrand** [comment]
Great news! :D

**Comment by nfelger** [comment]
Thanks for being so persistent! I just had the same problem, and going from the ARMv6 library to the ARMv5 version fixed it for me, too.

**Comment by Burgestrand** [comment]
Thanks for the confirmation!
